refactor(extension): [LW-9094] use new wallet manager

mkazlauskas opened the pull request:
In order to support multi-wallet/account & shared wallets, walletManager was completely redesigned in SDK.

There are a couple new components:
- WalletRepository stores key agent data, used accounts, as well as any additional metadata such as 'name'. Lace should use it to get an array of all users wallets.
- SignerManager is an abstraction over KeyAgent. Lace no longer needs to maintain the lifecycle of a KeyAgent. Instead, it should use this component to sign transactions.

For more info, see [cardano-js-sdk documentation](https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/web-extension/README.md)

There are also some other changes due to SDK update:
- Transaction inspectors are now async

# Checklist

- [x] JIRA LW-9094 LW-9533
- [ ] Proper tests implemented
- [ ] Screenshots added.

---

## Proposed solution

Explain how does this PR solves the problem stated in JIRA ticket.
You can also enumerate different alternatives considered while approaching this task.

## Testing

Describe here, how the new implementation can be tested.
Provide link or briefly describe User Acceptance Criteria/Tests that need to be met

## Screenshots

Attach screenshots here if implementation involves some UI changes

mkazlauskas committed:
ci: set higher memory limit for running ci unit tests

mkazlauskas committed:
test: update e2e suite to use wallet repository instead of localstorage directly

mkazlauskas committed:
test(extension): increase e2e timeout duration

mkazlauskas committed:
ci: set mainnet cardano-services url to dev-perf-mainnet

mkazlauskas committed:
chore: disable mangling of class names

code depends on class name being the same in sw and ui

mkazlauskas committed:
refactor(extension): [LW-9094] use new wallet manager

In order to support multi-wallet/account & shared wallets,
walletManager was completely redesigned in SDK.

There are a couple new components:
- WalletRepository stores key agent data, used accounts,
as well as any additional metadata such as 'name'. Lace
should use it to get an array of all users wallets.
- SignerManager is an abstraction over KeyAgent. Lace
now longer needs to maintain the lifecycle of a KeyAgent.
Instead, it should use this component to sign transactions.

There are also some other changes due to SDK update:
- Transaction inspectors are now async

rhyslbw reviewed (commented):
Looks good from a high level, although coveralls is reporting a drop of 3.5% test coverage which is out of the allowed threshold. We'll also need to run full regression testing on this PR @ljagiela to have full confidence, given it's a huge refactor.

DominikGuzei commented on apps/browser-extension-wallet/src/features/dapp/components/ConfirmTransaction.tsx:
```diff
@@ -192,17 +164,34 @@ export const ConfirmTransaction = withAddressBookContext((): React.ReactElement
-  }, []);
+    console.warn('e2e-debug', 'subscribe');
+    const subscription = signingCoordinator.transactionWitnessRequest$.subscribe(async (r) => {
+      console.warn('e2e-debug', 'got req', r);
```
do we want to keep these `console.log` statements?

mkazlauskas replied:
Nice catch. I'll leave the console.error ones unless there's a better way to log in lace?

mkazlauskas replied:
Removed debug logs in 02acf86e4f3ffeb311fdf43518fad755fd055e5d

DominikGuzei commented on apps/browser-extension-wallet/src/features/dapp/components/ConfirmTransaction.tsx:
```diff
@@ -101,10 +92,10 @@ export const ConfirmTransaction = withAddressBookContext((): React.ReactElement
+  const redirectToSignSuccess = useRedirection(dAppRoutePaths.dappTxSignSuccess);
   const [isConfirmingTx, setIsConfirmingTx] = useState<boolean>();
   const keyAgentType = getKeyAgentType();
   const isUsingHardwareWallet = useMemo(
```
Are we still using key agents? I thought this was replaced by the new `SigningCoordinator` – or is this something else?

mkazlauskas replied:
Nice catch, I remember wanting to refactor this but then forgot about it. Done in e72edfebcbab3cd3d6daf271d7b7852724cd61c6

DominikGuzei commented on apps/browser-extension-wallet/src/features/dapp/components/ConfirmTransaction.tsx:
```diff
@@ -249,57 +238,66 @@ export const ConfirmTransaction = withAddressBookContext((): React.ReactElement
-      type: txType,
-      mintedAssets: createMintedList(minted),
-      burnedAssets: createMintedList(burned)
+      // eslint-disable-next-line unicorn/no-array-reduce
```
Maybe we should disable `unicorn/no-array-reduce` if we still want to use it?

mkazlauskas replied:
Agree 6e71227b31d86581e55e022eafd4f1ed37c9e3f1

DominikGuzei commented on apps/browser-extension-wallet/src/features/dapp/components/confirm-transaction/utils.ts:
```diff
@@ -0,0 +1,157 @@
+  return sectionTitle[DAPP_VIEWS.CONFIRM_TX];
+};
+
+export const disallowSignTx = (close = false): void => {
```
It would be great to document this function as I don't understand what it does exactly

mkazlauskas replied:
This one wasn't used, removed in 688280f8b9b6277c399e310042e33a903486d040

DominikGuzei commented on apps/browser-extension-wallet/src/features/dapp/components/confirm-transaction/utils.ts:
```diff
@@ -0,0 +1,157 @@
+  close && setTimeout(() => window.close(), DAPP_TOAST_DURATION);
+};
+
+export const allowSignTx = (): void => {
```
It would be great to document this function with a comment as I don't understand what it does exactly

mkazlauskas replied:
Renamed for better clarity and added a doc comment in 688280f8b9b6277c399e310042e33a903486d040

DominikGuzei commented on apps/browser-extension-wallet/src/features/unlock-wallet/components/UnlockWalletContainer.tsx:
```diff
@@ -84,13 +68,23 @@ export const UnlockWalletContainer = ({ validateMnemonic }: UnlockWalletContaine
-      onValidationSuccess={loadWallet}
-    />
-  ) : (
+  // LW-9500
```
Do we want to keep this commented-out code?

mkazlauskas replied:
Agree, removed in d31c1fd508d2ad4464bf660bf0e25e84e865f616 and added a note in ticket description that it can be found in git before new wallet manager

mkazlauskas committed:
fix(extension): show own address when confirming SignData

mkazlauskas commented on apps/browser-extension-wallet/src/lib/scripts/background/services/expose.ts:
```diff
@@ -0,0 +1,19 @@
+import { of } from 'rxjs';
+import { runtime } from 'webextension-polyfill';
+
+// This was hoisted from userIdService.ts so that it's not being exposed while running the unit tests of the class itself.
```
LW-9667

mkazlauskas committed:
chore(extension): remove debug console logs

lucas-barros commented on apps/browser-extension-wallet/src/hooks/useWalletManager.ts:
```diff
@@ -154,162 +299,182 @@ export const useWalletManager = (): UseWalletManager => {
+          ? HexBlob.fromBytes(lock)
+          : undefined;
+      // lock behavior changed a little bit, it is only in local storage while actually locked
+      // REVIEW: it is likely that Lace update/migration would unlock a locked wallet. Is that ok?
```
You could check if the lock is active and update the lock value?

mkazlauskas replied:
Done in 90c33fa05136778db3617270a6bec27754b58d1e
Let's make sure it's tested in QA

lucas-barros commented on apps/browser-extension-wallet/src/lib/wallet-api-ui.ts:
```diff
@@ -0,0 +1,86 @@
+  try {
+    return await action();
+  } finally {
+    subscription.unsubscribe();
```
Could we clear the password here?

mkazlauskas replied:
Good idea, done in 741b3994c6f8008a80823da01089764f68683bd8
Although it would be better if it came as byte array already as that we can actually overwrite in memory

lucas-barros commented on apps/browser-extension-wallet/src/views/browser-view/features/staking/hooks.ts:
```diff
@@ -1,14 +1,26 @@
+    const tx = delegationTxBuilder.build();
+    const { hash } = await tx.inspect();
+    void firstValueFrom(
+      cardanoWallet.signingCoordinator.transactionWitnessRequest$.pipe(
```
How does this work? Could we use withSignTxConfirmation here?

mkazlauskas replied:
Nice catch, this is just some code duplication I left. No need to actually check tx hash here. Done in 38cf7bf51d59d01f5fc6e500dbc64aeaab1d5be7

mkazlauskas committed:
refactor(extension): replace keyAgentType with walletType

mkazlauskas committed:
chore: disable eslint unicorn/no-array-reduce rule

mkazlauskas committed:
refactor(extension): rename allowSignTx to readyToSignData

also remove an obsolete disallowSignTx function

mkazlauskas committed:
chore(extension): remove commented out code for mnemonic verification

mkazlauskas committed:
chore: rollback noEmitOnError tsconfig to true

mkazlauskas commented on apps/browser-extension-wallet/src/features/dapp/components/ConfirmData.tsx:
```diff
@@ -59,79 +60,69 @@ export const DappConfirmData = (): React.ReactElement => {
-      });
-  }, []);
+    const dataFromHex = fromHex(req.blob);
+    // TODO: address is currently not available in sign request,
```
Yes, address is displayed in sign data confirmation page. It's available now, fixed in 6c1a1b013b51f585a50023fdebe4a3c03a1692ca

mkazlauskas commented on apps/browser-extension-wallet/src/features/dapp/components/ConfirmData.tsx:
```diff
@@ -59,79 +60,69 @@ export const DappConfirmData = (): React.ReactElement => {
+    const api = exposeApi<Pick<UserPromptService, 'allowSignTx'>>(
       {
         api$: of({
-          async allowSignData(): Promise<boolean> {
```
It is, nice catch. Fixed this in the rename commit here 688280f8b9b6277c399e310042e33a903486d040

mkazlauskas commented on tsconfig.json:
```diff
@@ -9,7 +9,7 @@
     "declaration": true,
     "declarationMap": true,
-    "noEmitOnError": true,
+    "noEmitOnError": false,
```
Sounds right, removed this change in 24c957686f7e3c10f2e144d37c67d48f4a7ec21c

mkazlauskas commented on packages/cardano/src/wallet/lib/cardano-wallet.ts:
```diff
@@ -269,63 +123,35 @@ export const validateWalletPassword = async (
       mnemonicWords,
-      chainId: originalKeyAgent.chainId,
-      getPassphrase: getPassword
+      chainId: Cardano.ChainIds.Mainnet, // does not matter
```
This function compares extended account public key: the one we have with the one we derive from mnemonic. It is the same for all networks, but we're deriving it by creating a key agent which has `chainId` as a required parameter. Ideally we would have a util that just derives the xpub without the redundant parameters, but working with what we have I think it's better if it we set it to a valid value instead of omitting and explicit type cast. Do you think otherwise?

mkazlauskas committed:
fix(extension): migrate locked wallet

mkazlauskas committed:
fix(extension): dereference password string after signing

mkazlauskas committed:
refactor(extension): utilize withSignTxConfirmation for useDelegationTransaction

mkazlauskas committed:
Merge remote-tracking branch 'origin/main' into LW-9094-new-wallet-manager

mkazlauskas committed:
chore(extension): remove obsolete shutdownWallet function

also remove some review comments

and add doc comments about how validateWalletMnemonic works

mchappell commented on apps/browser-extension-wallet/src/features/dapp/components/ConfirmData.tsx:
```diff
@@ -40,98 +40,84 @@ const hasJsonStructure = (str: string): boolean => {
-  const cancelTransaction = useCallback(() => {
-    exposeApi<Pick<UserPromptService, 'allowSignData'>>(
+  const cancelTransaction = useCallback(async () => {
+    await req.reject('User rejected to sign');
```
Same comment here, should we be rejecting with `ApiError` and `APIErrorCode.Refused` or does the sdk spit that out?

mkazlauskas replied:
SDK wraps it in `AuthenticationError` (from key-management), which should then be wrapped in `TxSignError` in cip30.ts :thinking:

mchappell commented on apps/browser-extension-wallet/src/features/dapp/components/ConfirmTransaction.tsx:
```diff
@@ -145,64 +134,55 @@ export const ConfirmTransaction = withAddressBookContext((): React.ReactElement
-  }, []);
+  const cancelTransaction = useCallback(
+    async (close = false) => {
+      await req.reject('User rejected to sign');
```
Same comment, if we reject here, it'll drop into the catch statement of `cip30.ts` won't it, which will return unable to sign, rather than an explicit rejection code.

mchappell commented on apps/browser-extension-wallet/src/stores/slices/wallet-info-slice.ts:
```diff
@@ -11,23 +10,21 @@ export const walletInfoSlice: SliceCreator<WalletInfoSlice & BlockchainProviderS
     get().setBlockchainProvider(chain);
   },
-  getKeyAgentType: () => get()?.cardanoWallet?.keyAgent.serializableData.__typename,
+  getWalletType: () => get()?.cardanoWallet?.source.wallet.type,
```
if we also added a variable for this rather than a function, we'd negate the need to perform this check in so many places and just use the type directly

mkazlauskas replied:
Done in 9f36f17a21f5a090465eb66a616f34d1348a1064

mkazlauskas committed:
fix(extension): fix wallet data migration to repository

broken in 90c33fa0

mkazlauskas committed:
Revert "ci: set mainnet cardano-services url to dev-perf-mainnet"

This reverts commit c611181811c9876629ee3dfd5ffb03e49d2f1043.

mkazlauskas committed:
fix(extension): log hd wallet in analytics only when it has addr with index >0

mkazlauskas committed:
refactor(staking): make walletManagerExecuteWithPassword generic again

DominikGuzei commented on apps/browser-extension-wallet/src/features/dapp/components/confirm-transaction/utils.ts:
```diff
@@ -0,0 +1,146 @@
+    {
+      api$: of({
+        async readyToSignTx(): Promise<boolean> {
+          return Promise.resolve(true);
```
I guess I don't understand how just resolving a boolean value here can be of any use? Ok there is one scenario i can think of: the service worker is polling this `readyToSignTx` endpoint and when the JS in the UI has initialized it returns true?

mkazlauskas replied:
Did you see 688280f8b9b6277c399e310042e33a903486d040 ? I added a doc comment for `readyToSignTx`.
It's basically this:

> the service worker is polling this readyToSignTx endpoint and when the JS in the UI has initialized it returns true

Except that it doesn't have to do any polling, it just waits for extension messaging connection to establish, then this is called and resolves, notifying the service worker that UI is ready to sign.

DominikGuzei replied:
Ok, after looking at the `exposeApi` + `api$` + `of` in the SDK, is it true that this creates an observable from the `allowSignTx` api method which can then be observed by the remote (service worker)?

mkazlauskas replied:
`api$` can be used to switch the exposed object (like the ObservableWallet that is exposed from service worker). In this case it uses `of` which just emits single value and completes. So this is not utilizing this capability, just exposing a specific api object. This is not new a new setup btw, just renamed the method from `allowSignTx` and changed it's purpose a little bit so that it's doing less - only used for coordination to notify service worker when UI is initialized.

DominikGuzei replied:
thanks for the explanation 👍 yes I have read the doc comment and for me it was still not obvious what's going on, so after your explanation maybe let's change it to:
```
// Exposes an `allowSignTx` observable, which emits a single `true` value to the service worker to indicate that the UI has loaded and is ready to sign
```

what do you think?

mkazlauskas replied:
Updated comment in cd41d9e5706f6455558f2e5bd27df9a179938277

DominikGuzei commented on apps/browser-extension-wallet/src/lib/scripts/background/wallet.ts:
```diff
@@ -1,44 +1,87 @@
+// It is important that this file is not exported from index,
+// because creating wallet repository with store creates an actual pouchdb database
+// which results in some trash files when running the tests (leveldb directory)
+export const walletRepository = new WalletRepository({
```
hm, should we then create the wallet repository instance in the root of the file?

mkazlauskas replied:
Could you please elaborate on this suggestion? :thinking:

DominikGuzei replied:
@mkazlauskas your comment suggests that creating the pouchdb database directly in the root of the file complicates the handling of this file, so I wondered if it might be possible to defer the creation with a setup function and call it separately?

mkazlauskas replied:
It could be deferred by converting this into `getWalletRepository` (as well as any other exports that depends on this one), but I personally prefer this for cleaner code. Ideally this would be created in index script (e.g. background.ts) and passed as dependency everywhere else, but that's a bigger refactor that I don't think we should do in this PR.

mkazlauskas committed:
refactor(extension): convert getWalletType into properties in wallet slice

walletType, isInMemoryWallet and isHardwareWallet

in order to make code in the components cleaner, as it's a common operation

mkazlauskas commented on apps/browser-extension-wallet/src/features/dapp/components/confirm-transaction/utils.ts:
```diff
@@ -0,0 +1,157 @@
+    {
+      api$: of({
+        async allowSignTx(): Promise<boolean> {
+          return Promise.reject(new Error('Transaction rejected'));
```
This was removed

mkazlauskas commented on apps/browser-extension-wallet/src/features/dapp/components/ConfirmData.tsx:
```diff
@@ -40,98 +40,84 @@ const hasJsonStructure = (str: string): boolean => {
-        });
-    } catch {
+      if (req.walletType !== WalletType.Ledger && req.walletType !== WalletType.Trezor) {
+        throw new Error('Invalid state: expected hw wallet');
```
I'm not sure: this is an assertion for state that should never happen. It doesn't mean that user rejected a transaction, but that there's a bug in Lace

mkazlauskas committed:
docs(extension): improve doc comment for readyToSign function

mkazlauskas commented on .github/shared/build/action.yml:
```diff
@@ -8,6 +8,9 @@ inputs:
   LACE_EXTENSION_KEY:
     description: 'Public extended manifest key'
     required: true
+  CARDANO_SERVICES_URL_MAINNET:
```
Reverted in 7cd2b62b0662443ee8b21117ab76b2076e518f9c

mkazlauskas commented on .github/workflows/smoke-tests.yml:
```diff
@@ -26,6 +26,7 @@ jobs:
         uses: ./.github/shared/build
         with:
           LACE_EXTENSION_KEY: ${{ secrets.MANIFEST_PUBLIC_KEY }}
+          CARDANO_SERVICES_URL_MAINNET: https://backend.dev-perf-mainnet.eks.lw.iog.io
```
Reverted in 7cd2b62b0662443ee8b21117ab76b2076e518f9c

mkazlauskas commented on apps/browser-extension-wallet/src/views/browser-view/features/wallet-setup/components/HardwareWalletFlow.tsx:
```diff
@@ -175,14 +175,7 @@ export const HardwareWalletFlow = ({
-        }
-        addressDiscoverySubscriber.unsubscribe();
-      });
+      analytics.sendEventToPostHog(PostHogAction.OnboardingRestoreHdWallet);
```
Nice catch! This was a bug, fixed in c68ccbe97560198f238d6b716e04bbba3e13c91d

Note I hoisted it one level up to WalletSetupWizard, because wallet has to be activated in order for `addresses$` to emit.

mchappell replied:
yeh that was likely my fault when we did the review on the other SDK pr, I spotted it then put it back but I think you'd already done the merge at that point

mkazlauskas commented on packages/staking/src/features/outside-handles-provider/types.ts:
```diff
@@ -52,6 +52,7 @@ export interface IBlockchainProvider {
 
 export type OutsideHandlesContextValue = {
   analytics: IAnalyticsTracker;
+  walletManagerExecuteWithPassword: (action: () => Promise<unknown>, password?: string) => Promise<unknown>;
```
No reason, I deleted it during development and then re-added with unknown as (I believe) return type wasn't used anywhere. Fixed in ad868dccb2736c690645d1a95d1e5598816891df

mkazlauskas commented on packages/cardano/src/wallet/lib/hardware-wallet.ts:
```diff
@@ -1,14 +1,9 @@
-import * as Crypto from '@cardano-sdk/crypto';
+import * as KeyManagement from '@cardano-sdk/key-management';
+import { DeviceConnection, HardwareWallets } from '../types';
 import * as HardwareLedger from '../../../../../node_modules/@cardano-sdk/hardware-ledger/dist/cjs';
```
I tried, but unfortunately it's a rabbit hole. I can try to spend a couple more hours on this separately, after this is merged.

mkazlauskas commented on apps/browser-extension-wallet/src/components/MainMenu/DropdownMenuOverlay/components/LockWallet.tsx:
```diff
@@ -1,17 +1,22 @@
-  const { walletLock } = useWalletStore();
+  const { getKeyAgentType } = useWalletStore();
   const analytics = useAnalyticsContext();
+  const isInMemoryWallet = useMemo(
```
Done in 9f36f17a21f5a090465eb66a616f34d1348a1064

Note: didn't add isInMemoryWallet and isHardwareWallet props for `staking` package state, as it's only used in a couple places.

DominikGuzei reviewed (approved):
great work 🎉

mkazlauskas committed:
fix(extension): do not overwrite tx/data being signed when a new request comes in

mkazlauskas committed:
Merge remote-tracking branch 'origin/main' into LW-9094-new-wallet-manager

mkazlauskas committed:
chore: bump sdk peerDependencies in staking package

mkazlauskas committed:
Merge remote-tracking branch 'origin/main' into LW-9094-new-wallet-manager

mkazlauskas committed:
refactor: store extendedAccountPublicKey on Bip32WalletAccount instead of Bip32Wallet

also update cardano-js-sdk packages

mkazlauskas committed:
refactor(extension): reload wallet when wallets$ or activeWalletId$ emits

Since wallet is now activated automatically,
add new setStayOnAllDonePage store property

mkazlauskas committed:
Merge remote-tracking branch 'origin/main' into LW-9094-new-wallet-manager

mkazlauskas committed:
fix(extension): update sdk to properly deactivate wallet

mkazlauskas committed:
test: restore order of closeAllTabsExceptOriginalOne in e2e after hook

mkazlauskas committed:
fix(extension): unlock wallet after migration to wallet manager

VanessaPC committed:
fix: dialogue title and description to be white on dark ui (#872)

mkazlauskas committed:
test: switch to Lace window before clearing repository

vetalcore committed:
chore(extension): remove circular deps, fix antd warnings (#861)

mkazlauskas committed:
fix(extension): do not reactivate wallet that's being deleted

Deleting a wallet is a 2 step process: deactivate and remove from
repository. When wallet is deactivated, `loadWallet` would immediately
reactivate it, leaving WalletManager and WalletRepository in an
incosistent state

mkazlauskas committed:
refactor: rm obsolete background service password cache utils

mkazlauskas committed:
test(extension): fix unit tests broken in 24bb2d39

mkazlauskas committed:
fix(extension): register posthog user id when active wallet changes

e2e suite is initializing local storage before the test.
When service worker starts, wallet repository (and active wallet)
is not yet initialized, so it would register a random user id
and never switch to wallet-based user id. This was only an issue
after localstorage->repository migration.

mkazlauskas committed:
fix: update local storage assertions to use wallet repository

mchappell committed:
chore: implement new cluster (#876)

ljagiela committed:
test(extension): e2e test adjustments

mkazlauskas committed:
fix(extension): register posthog distinct_id only after init resolves

mkazlauskas committed:
fix(extension): display the signed payload rather than entire cip8 structure

ljagiela committed:
test(extension): fix for LW-4677 e2e test

mkazlauskas committed:
fix(extension): disable analytics when it fails to initialize

ljagiela committed:
test(extension): fix for @LW-9279 e2e test

mkazlauskas committed:
fix(extension): rename transaction pouchdb databases

hoist createPouchdbWalletStores code from sdk

in order to rename transaction databases due to a breaking change in tx obj shape

mkazlauskas committed:
fix(extension): restore analytics behavior of sending session start events

mkazlauskas committed:
Merge remote-tracking branch 'origin/main' into LW-9094-new-wallet-manager

mkazlauskas committed:
refactor(extension): add current wallet state into store at wallet info slice

Subscribe to all wallet observables at app init and put it as plain
object into store. Then use it for computing wallet activities.

This refactor was made due to performance degradation when loading
Activity page for large wallets.

mkazlauskas committed:
Merge remote-tracking branch 'origin/main' into LW-9094-new-wallet-manager

mkazlauskas committed:
chore(extension): remove matomo

there are still some mentions in privacy/cookie policy left

mkazlauskas committed:
chore: bump @cardano-sdk packages

VanessaPC committed:
fix: remove ADA_HANDLE feature flag leftovers

mkazlauskas committed:
Merge remote-tracking branch 'origin/main' into LW-9094-new-wallet-manager

mkazlauskas committed:
fix: guard against undefined walletState in useWalletActivities

mkazlauskas committed:
Revert "fix: remove ADA_HANDLE feature flag leftovers"

This reverts commit a8b1b8b8ad7da1022e10d3ef6cd40cadd2fdbb5e.

mkazlauskas committed:
fix(extension): preserve selected network after forgot pw flow

mkazlauskas committed:
fix(extension): preserve selected network when migrating locked wallet

ljagiela committed:
test(extension): e2e adjustments

mkazlauskas committed:
Merge remote-tracking branch 'origin/main' into LW-9094-new-wallet-manager

ljagiela committed:
test(extension): e2e adjustments

ljagiela committed:
test(extension): e2e remove check for localStorage appSettings

mkazlauskas committed:
refactor: replace CML with @cardano-sdk/crypto implementation

mkazlauskas committed:
feat: convert handle to lowercase before validating/resolving

handles are not case sensitive

also update @cardano-sdk packages

mkazlauskas committed:
Merge remote-tracking branch 'origin/main' into LW-9094-new-wallet-manager

mkazlauskas committed:
Merge remote-tracking branch 'origin/main' into LW-9094-new-wallet-manager

mchappell commented on packages/cardano/src/wallet/test/mocks/TestKeyAgent.ts:
```diff
@@ -23,6 +21,6 @@ export const testKeyAgent = ({
     {
       logger: console,
-      bip32Ed25519: new Crypto.CmlBip32Ed25519(CML)
+      bip32Ed25519: new Crypto.SodiumBip32Ed25519()
```
🚀

mkazlauskas committed:
Merge remote-tracking branch 'origin/main' into LW-9094-new-wallet-manager

mchappell reviewed (commented):
LGTM

mkazlauskas commented on apps/browser-extension-wallet/src/views/browser-view/features/send-transaction/components/Form/Form.tsx:
```diff
@@ -159,13 +152,6 @@ export const Form = ({
           fiatAmount={totalCost.fiat}
           tooltipContent={t('send.theAmountYoullBeChargedToProcessYourTransaction')}
-          onTooltipHover={() =>
-            analytics.sendEventToMatomo({
```
I don't know, but it wasn't, so it should probably be out of scope for this PR

mkazlauskas committed:
chore: cleanup

mkazlauskas commented on apps/browser-extension-wallet/src/views/browser-view/features/staking/components/StakePoolDetails/StakePoolDetailsDrawer.tsx:
```diff
@@ -39,22 +37,19 @@ export const StakePoolDetailsDrawer = ({
+  const { isInMemoryWallet } = useWalletStore();
   const { password, removePassword } = usePassword();
-  const isInMemory = useMemo(() => getKeyAgentType() === Wallet.KeyManagement.KeyAgentType.InMemory, [getKeyAgentType]);
   // const isSuccessSection = useMemo(
```
https://github.com/input-output-hk/lace/pull/816/commits/cc4fa8e6d7b9967bc1a44ec8dbd6f733c09de4d1

VanessaPC commented on apps/browser-extension-wallet/src/features/dapp/components/ConfirmData.tsx:
```diff
@@ -40,105 +40,90 @@ const hasJsonStructure = (str: string): boolean => {
+    setTimeout(() => window.close(), DAPP_TOAST_DURATION);
+  }, [req]);
+
+  window.addEventListener('beforeunload', cancelTransaction);
```
generally adding add event listeners but not removing them is discouraged, I would suggest: 
```
  useEffect(() => {
    window.addEventListener('beforeunload', cancelTransaction);

    return () => {
      window.removeEventListener('beforeunload', cancelTransaction);
    };
  }, [cancelTransaction]);
  ```

mkazlauskas replied:
Agree, but I tried to avoid making changes unless something got in the way of wallet manager refactor. This was already like that: https://github.com/input-output-hk/lace/pull/816/files#diff-36b0722aea432cfca18257c06db391c57e9d8a0d8fbd9dd9281237b45fe91936L78 Let me know if you want me to resolve this issue in this PR or if it's ok to keep as-is.

VanessaPC replied:
I thought it was a change from the PR, let's keep it as is.

VanessaPC commented on apps/browser-extension-wallet/src/features/dapp/components/collateral/CreateCollateral.tsx:
```diff
@@ -65,39 +62,36 @@ export const CreateCollateral = ({
+    try {
+      await withSignTxConfirmation(submitTx, password);
+    } catch (error) {
+      if (error instanceof Wallet.KeyManagement.errors.AuthenticationError) {
```
nice

VanessaPC reviewed (approved):
I left a small comment, it looks good to me!
